Print friendly error when `release create` fails due to missing `workflow` OAuth scope

BagToad opened the pull request:
Resolves https://github.com/cli/cli/issues/9773 by adding a friendly error message when `gh release create` fails and it may be due to a missing `workflow` OAuth scope.

Example leveraging a fork of the repo in https://github.com/cli/cli/issues/9773:

```
gh release create v0.1.0 --target a602187ebae72cc0e8101cc46fc116a51c9e9e39 --title foo --notes bar -R bagtoad/bevy-yoetz
HTTP 404: Failed to create release, "workflow" scope may be required
To request it, run gh auth refresh -h github.com -s workflow
```

BagToad committed:
Handle missing "workflow" scope in createRelease

andyfeller commented:
@BagToad : **Is there anything you need to go from a draft to a full PR?**

I wasn't sure if you created a draft for discussion or feedback purposes.  Just making sure to understand what's needed here. 👍

BagToad commented on pkg/cmd/release/create/http.go:
```diff
@@ -174,6 +178,20 @@ func createRelease(httpClient *http.Client, repo ghrepo.Interface, params map[st
+	// if the same file (with both the same path and contents) exists
+	// on another branch in the repo. Otherwise, the workflow scope is required.
+	oAuthTokenMissingWorkflowScope := isOAuthToken(resp) && !oAuthTokenHasWorkflowScope(resp)
+	if resp.StatusCode == 404 && oAuthTokenMissingWorkflowScope {
```
Note: need to use HTTP constant here instead

BagToad commented on pkg/cmd/release/create/http.go:
```diff
@@ -174,6 +178,20 @@ func createRelease(httpClient *http.Client, repo ghrepo.Interface, params map[st
 	}
 	defer resp.Body.Close()
 
+	// Releases with workflow files can be created without the workflow scope,
+	// if the same file (with both the same path and contents) exists
+	// on another branch in the repo. Otherwise, the workflow scope is required.
+	oAuthTokenMissingWorkflowScope := isOAuthToken(resp) && !oAuthTokenHasWorkflowScope(resp)
+	if resp.StatusCode == 404 && oAuthTokenMissingWorkflowScope {
+		normalizedHostname := ghauth.NormalizeHostname(resp.Request.URL.Hostname())
+		errMissingRequiredWorkflowScope := errors.New(heredoc.Docf(`
+				HTTP 404: Failed to create release, "workflow" scope may be required
+				To request it, run gh auth refresh -h %[1]s -s workflow
+			`, normalizedHostname))
+
+		return nil, errMissingRequiredWorkflowScope
+	}
+
```
I'm not a fan of how this is written and am looking for feedback on ways to improve it.

andyfeller replied:
Honestly, I'm torn here and left https://github.com/cli/cli/issues/9773#issuecomment-2432134253 to explain.  I think this might be an opportunity to try improving the REST API endpoint behavior rather than trying to work around it in `gh`.

Beyond that, here are other places in code where we handle similar situations:

https://github.com/cli/cli/blob/e390874fef5eb5545db2c4ff0861cb1c3e83dd6f/pkg/cmd/auth/setupgit/setupgit.go#L87-L100

https://github.com/cli/cli/blob/e390874fef5eb5545db2c4ff0861cb1c3e83dd6f/pkg/cmd/gpg-key/add/add.go#L84-L105

https://github.com/cli/cli/blob/e390874fef5eb5545db2c4ff0861cb1c3e83dd6f/pkg/cmd/project/shared/queries/queries.go#L1458-L1466

BagToad commented on pkg/cmd/release/create/http.go:
```diff
@@ -254,3 +272,20 @@ func deleteRelease(httpClient *http.Client, release *shared.Release) error {
 	}
 	return nil
 }
+
+func isOAuthToken(resp *http.Response) bool {
+	scopes := resp.Header.Get("X-Oauth-Scopes")
+	return scopes != ""
+}
+
+func oAuthTokenHasWorkflowScope(resp *http.Response) bool {
+	scopes := resp.Header.Get("X-Oauth-Scopes")
+
+	for _, s := range strings.Split(scopes, ",") {
+		if s == "workflow" {
+			return true
+		}
+	}
+
+	return false
+}
```
I would have loved for these functions to live somewhere else instead of locally in this command. I'm also not a fan of having two functions here. Open to feedback on ways to improve this situation.

andyfeller commented on pkg/cmd/release/create/http.go:
```diff
@@ -174,6 +178,26 @@ func createRelease(httpClient *http.Client, repo ghrepo.Interface, params map[st
+	// required to create the release.
+	//
+	// https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/scopes-for-oauth-apps#available-scopes
+	if resp.StatusCode == http.StatusNotFound && tokenMissingWorkflowScope(resp) {
+		normalizedHostname := ghauth.NormalizeHostname(resp.Request.URL.Hostname())
+		errMissingRequiredWorkflowScope := errors.New(heredoc.Docf(`
+				HTTP 404: Failed to create release, "workflow" scope may be required
+				To request it, run gh auth refresh -h %[1]s -s workflow
+			`, normalizedHostname))
+
+		return nil, errMissingRequiredWorkflowScope
+	}
```
This technically works though I don't know how much this should align to other places we  have to do similar errors:

https://github.com/cli/cli/blob/1ec23f698f50f709a54099aba633d0031d21d6b4/pkg/cmd/gpg-key/add/add.go#L87-L91

https://github.com/cli/cli/blob/1ec23f698f50f709a54099aba633d0031d21d6b4/pkg/cmd/gpg-key/list/list.go#L60-L65

https://github.com/cli/cli/blob/1ec23f698f50f709a54099aba633d0031d21d6b4/pkg/cmd/project/shared/queries/queries.go#L1461-L1466

https://github.com/cli/cli/blob/1ec23f698f50f709a54099aba633d0031d21d6b4/pkg/cmd/auth/status/status.go#L47-L54

To summarize:

1. Should we print the error here and return a silent error?
2. Should we emphasize the command to run?
3. Should we avoid the `HTTP 404` in the error message?

🤔

BagToad replied:
> Should we print the error here and return a silent error?

What is the benefit for returning this silent error?

> Should we emphasize the command to run?

Sure, I like that idea.

> Should we avoid the HTTP 404 in the error message?

Yes, I can remove it.

BagToad replied:
> > Should we print the error here and return a silent error?
> 
> What is the benefit for returning this silent error?

Of course, the benefit is that we can print the error right here instead of passing an `error` which is eventually printed 🤪 

> > Should we emphasize the command to run?
> 
> Sure, I like that idea.
> > Should we avoid the HTTP 404 in the error message?
> 
> Yes, I can remove it.

Done 👍 

----

I made these changes and a refactor in deb34d6456d14efe73ccd61fcb88e063da5f4794

Also, thank you for the examples @andyfeller ✨ - they were very helpful.

andyfeller commented on pkg/httpmock/stub.go:
```diff
@@ -238,6 +238,20 @@ func ScopesResponder(scopes string) func(*http.Request) (*http.Response, error)
 	}
 }
 
+// StatusScopesResponder returns a response with the given status code and OAuth scopes.
+func StatusScopesResponder(status int, scopes string) func(*http.Request) (*http.Response, error) {
+	return func(req *http.Request) (*http.Response, error) {
+		return &http.Response{
+			StatusCode: status,
+			Request:    req,
+			Header: map[string][]string{
+				"X-Oauth-Scopes": {scopes},
+			},
+			Body: io.NopCloser(bytes.NewBufferString("")),
+		}, nil
+	}
+}
```
Given how similar this is to `ScopesResponder`, should this go away and that function be updated to require providing the status code?  Or have it be a wrapper to this and passing the default status code?

My knee jerk is avoiding the wrapper and have callers' specify the exepected status code. 🤷

BagToad replied:
`ScopesResponder` is exported and used 35 times in commands unrelated to this PR. Personally, I'd prefer to avoid a signature change and thus needing to change those other tests, if possible.

![image](https://github.com/user-attachments/assets/589f95dd-0aa1-48dc-84a4-9083a51a298e)

What are your thoughts?

BagToad replied:
Updated `ScopesResponder` to wrap `StatusScopesResponder` in 11dc6df. I think if a refactor is required to the 35 existing tests through changing `ScopesResponder`'s signature, it should be done in another PR and shouldn't block the value in this PR from being released.

andyfeller commented on pkg/cmd/release/create/http.go:
```diff
@@ -254,3 +278,24 @@ func deleteRelease(httpClient *http.Client, release *shared.Release) error {
+		return false
+	}
+
+	for _, s := range strings.Split(scopes, ",") {
+		if s == "workflow" {
+			return false
+		}
+	}
+
+	return true
```
I always struggle when we should do iteration like this versus something simpler like [`slices.Contains()`](https://pkg.go.dev/slices#Contains):

```suggestion

	return slices.Contains(strings.Split(scopes, ","), "workflow")
```

BagToad replied:
Good call, thanks for the refactor 😁

andyfeller commented on pkg/cmd/release/create/create_test.go:
```diff
@@ -1082,6 +1083,72 @@ func Test_createRun(t *testing.T) {
+			},
+			wantErr: "HTTP 404 (https://api.github.com/repos/OWNER/REPO/releases)",
+		},
+		{
+			name:  "API returns 404, not an OAuth token",
+			isTTY: false,
+			opts: CreateOptions{
+				TagName: "Does not matter",
+			},
+			runStubs: func(rs *run.CommandStubber) {
+				rs.Register(contentCmd, 0, "some tag message")
+				rs.Register(signatureCmd, 0, "")
+			},
+			httpStubs: func(t *testing.T, reg *httpmock.Registry) {
+				reg.Register(
+					httpmock.GraphQL("RepositoryFindRef"),
+					httpmock.StringResponse(`{"data":{"repository":{"ref": {"id": "tag id"}}}}`),
+				)
+				reg.Register(
+					httpmock.REST("POST", "repos/OWNER/REPO/releases"),
+					httpmock.StatusStringResponse(404, `HTTP 404 (https://api.github.com/repos/OWNER/REPO/releases)`))
+			},
+			wantErr: "HTTP 404 (https://api.github.com/repos/OWNER/REPO/releases)",
+		},
```
If I understand correctly, other types of tokens don't report scopes associated, right?

BagToad replied:
Yes, at least not in the `X-Oauth-Scopes` header.

andyfeller reviewed (commented):
Solid work! 💪 Some random thoughts and concerns

BagToad committed:
Refactor `workflow` scope checking

Refactor the logic for checking `workflow` scope checking in releases to be in the positive - check if the scope is there, not check if it isn't there. Then, when the function is called we invert it.

Also update comments to be more imperative.

This refactor also incorporates @andyfeller's suggestion to use `slices`.

Co-Authored-By: Andy Feller <andyfeller@github.com>

BagToad committed:
ScopesResponder wraps StatusScopesResponder

BagToad committed:
Refactor error handling for missing "workflow" scope in createRelease

BagToad commented on pkg/cmd/release/create/create_test.go:
```diff
@@ -1115,7 +1184,6 @@ func Test_createRun(t *testing.T) {
 			err := createRun(&tt.opts)
 			if tt.wantErr != "" {
 				require.EqualError(t, err, tt.wantErr)
-				return
 			} else {
 				require.NoError(t, err)
 			}
```
Not sure why this `return` was here, it seems like it might have been a mistake.

I needed to remove it to get my test to initially fail since my test wants a `SilentError` while also wanting text on stderr 🧐

andyfeller replied:
I think this `return` would cause tests that `wantErr` to skip checking `stdout` and `stderr`.  Some of the tests define empty `wantStdout` and `wantStderr`, so removing this `return` would still work for most:

### From `trunk`:

https://github.com/cli/cli/blob/ae87caf8d65f8d5a84646384a68641c0a4d25900/pkg/cmd/release/create/create_test.go#L833-L835

https://github.com/cli/cli/blob/ae87caf8d65f8d5a84646384a68641c0a4d25900/pkg/cmd/release/create/create_test.go#L1115-L1124

### From https://github.com/cli/cli/pull/1552:

You can see the test used to check the resulting release information, too:

https://github.com/cli/cli/blob/0f42c7af4d014b8d711241630099fadd1b588a79/pkg/cmd/release/create/create_test.go#L314-L330

andyfeller reviewed (approved):
The inline and function documentation is icing on the cake!
